.github/workflows/push.yml: add clang-tidy job

MaxKellermann opened the pull request:
As discussed on https://github.com/php/php-src/pull/10588

MaxKellermann commented:
Apparently, the `apt-x64` action doesn't work on Ubuntu 22.04, but I want that version so I can get a newer clang-tidy version.

devnexen commented:
you might need to filter out some things here and there, some warnings are noises but it s unavoidable, some of but it s good start !

MaxKellermann commented:
> you might need to filter out some things here and there

I already disabled a few that looked like you weren't interested, but I wasn't sure about others like `bugprone-macro-parentheses`. That warning is correct, and there should be parentheses, though that's a kludge and I'd prefer not using macros at all...

But if you have a list of checks that you think should be disabled, I'll amend the PR commit.

devnexen commented:
Well, for example little things like `readability-magic-numbers` is not exactly a real concerning issue I find.

MaxKellermann commented:
> `readability-magic-numbers` is not exactly a real concerning issue

Removed, and added a kludge to exclude failing generated sources (language_scanner).

devnexen commented:
ok will surely need more refinement but let s see what other folks think 🙂

nielsdos commented:
I see that Clang-tidy performs some of the same static analysis that clang-static-analyzer does. I ran clang-static-analyzer in the past on PHP and found it has a lot of trouble understanding the parameter parsing code, which leads to a huge number of false positive "uninitialized variable" warning. From a brief look it looks like this is the case here too, so that's something to be aware of.

mvorisek commented on .github/workflows/push.yml:
```diff
@@ -161,3 +161,22 @@ jobs:
             -d opcache.jit_buffer_size=16M
       - name: Verify generated files are up to date
         uses: ./.github/actions/verify-generated-files
+  TIDY:
```
```suggestion
  CLANG_TIDY:
```

mvorisek commented on .github/actions/configure-tidy/action.yml:
```diff
@@ -0,0 +1,37 @@
+name: ./configure
```
is special configure for clang-tidy needed? if not, and as clang-tidy is not the fastest, I would say regular apt/configure can be (re)used, for full analysis, all dev packages are needed probably anyway

MaxKellermann replied:
Not for clang-tidy, but for Ubuntu 22.04 - see my first comment at the top of the PR. My 22.04 versions of apt/configure disable lots of features (to sidestep the apt breakage) that could be re-enabled as soon as somebody figures that all out for 22.04 for the "regular" CI builds. What I found most important for now is that the main/Zend directories get checked by clang-tidy. When those code bases are "clean", we can go on to check all those extensions as well.

devnexen replied:
You might need to still filter out few readibility warnings especially case like

```
/home/runner/work/php-src/php-src/ext/date/lib/parse_date.c:12927:20: warning: statement should be inside braces [readability-braces-around-statements]
                        if (yych == 'y') goto yy810;
```

MaxKellermann replied:
But this warning is correct - this piece of code is wrong according to https://github.com/php/php-src/blob/master/CODING_STANDARDS.md

devnexen replied:
I m not saying it s wrong 🙂

1/ this particular case belongs to timelib I believe.
2/ Such warnings make it harder to see more "interesting cases".

But it is just a suggestion 😉

nielsdos replied:
The code highlighted is mostly generated code, so I don't think it makes sense to tidy that

MaxKellermann replied:
Yes, tool-generated code is out of your control; but it surprises me that generated code is committed to the git repository?
Most were actually removed 5 years ago by commit f0251a8b16631edf06830e08698cef5d76c1e01b but this one not.

petk replied:
@MaxKellermann some of these files were still Git-tracked because of this discussion https://externals.io/message/102796

I don't know if this has changed in the meantime.

The timelib code is maintained elsewhere and in php-src repo it is just bundled. It is bumped with its own release cycle on multiple PHP branches. It's written also in the contributing scheme https://github.com/php/php-src/blob/master/CONTRIBUTING.md#php-source-code-directory-structure

devnexen replied:
we might need, in some places, to filter out via `NOLINT` some macros (e.g. some Zend argument parse ones).

mvorisek commented:
Can clang-tidy pass if we disable all the rules?
